rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala

```diff
@@ -0,0 +1,375 @@
+package coop.rchain.rspace.nextgenrspace.history
+
+import coop.rchain.rspace.Blake2b256Hash
+import scodec.Codec
+import scodec.bits.{BitVector, ByteVector}
+import scodec.codecs.{discriminated, provide, uint2}
+import coop.rchain.rspace.internal.codecByteVector
+import coop.rchain.shared.AttemptOps._
+import History._
+import cats.{Applicative, FlatMap}
+import cats.effect.Sync
+import cats.implicits._
+sealed trait Trie
+sealed trait NonEmptyTrie                                   extends Trie
+case object EmptyTrie                                       extends Trie
+final case class Skip(affix: ByteVector, hash: TriePointer) extends NonEmptyTrie
+final case class Node(hash: PointerBlockPointer)            extends NonEmptyTrie
+final case class Leaf(hash: LeafPointer)                    extends NonEmptyTrie
+
+final case class TriePath(partialPath: Vector[Trie], conflicting: Option[Trie], path: KeyPath) {
+  def append(affix: KeyPath, t: Trie): TriePath =
```

Comment:
I'd make it:

```scala
TriePath(nodes: Vector[Trie], edges: KeyPath, conflictingSkip: Option[Skip]) {
```

Isn't there an invariant that `nodes.size == edges.size + 1`?

BTW why to we use a Vector for `nodes`? Do we need sparsity or random indexing? Wouldn't a List suffice?

Reply:
random indexing was the thought.

Reply:
edges is a bad name, it's a path. edges indicate that there is no relation between them.

Reply:
the more I think about it: nodes is also a bad name. partial path or partial prefix nodes is more precise.

Reply:
My problem is that it's unclear which is the labels, and which is the nodes. My proposal is to have path.nodes and path.edges :)

re: random indexing - I guess that's only needed because we operate on the end of the list instead of its head? That's not a big deal, but maybe worth considering a change in the future.

Reply:
but why? is vector a lesser collection?

Reply:
nodes and edges don't point to "nodes" and "labels"

sorry, I don't follow the need. it's a case class that has pieces of named and typed data. the names express their role. what's the gain of using less specific names?

Reply:
guess I'll never learn. My issue with the original names - `path` and `partialPath` is:
 - how is one partial, and the other isn't? Aren't they both 'partial'? (partial as in 'from the start to wherever we managed to get so far')
 - which one has the nodes on the path, and which one their indices in the PointerBlock?

With nodes + edges it's clear which is which. Partiality is kind of expected for me when we talk about paths. And if it applies to both paths, maybe let's have PartialTriePath or just PartialPath instead of TriePath.

Anyway, I voiced why this was causing problems in understanding. What to do with that is up to you.

Reply:
ok, I can see that the current names are not expressive enough. I don't agree that the proposed ones are better.